src/validators.ts

```diff
@@ -0,0 +1,135 @@
+import { Spec, ValidatorSpec } from './types'
+
+// Simplified adaptation of https://github.com/validatorjs/validator.js/blob/master/src/lib/isFQDN.js
+const isFQDN = (input: string) => {
+  if (!input.length) return false
+  const parts = input.split('.')
+  for (let part, i = 0; i < parts.length; i++) {
+    part = parts[i]
+    if (!/^[a-z\u00a1-\uffff0-9-]+$/i.test(part)) return false
+    if (/[\uff01-\uff5e]/.test(part)) return false // disallow full-width chars
+    if (part[0] === '-' || part[part.length - 1] === '-') return false
+  }
+  return true
+}
+
+// "best effort" regex-based IP address check
+// If you want a more exhaustive check, create your own custom validator, perhaps wrapping this
+// implementation (the source of the ipv4 regex below): https://github.com/validatorjs/validator.js/blob/master/src/lib/isIP.js
+const ipv4Regex = /^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$/
+const ipv6Regex = /([a-f0-9:]+:+)+[a-f0-9]+/
+const isIP = (input: string) => {
+  if (!input.length) return false
+  return ipv4Regex.test(input) || ipv6Regex.test(input)
+}
+
+const EMAIL_REGEX = /^[^@\s]+@[^@\s]+\.[^@\s]+$/ // intentionally non-exhaustive
+
+export class EnvError extends TypeError {
+  constructor(message?: string) {
+    super(message)
+    Error.captureStackTrace(this, EnvError)
+    this.name = 'EnvError'
+  }
+}
+
+export class EnvMissingError extends ReferenceError {
+  constructor(message?: string) {
+    super(message)
+    Error.captureStackTrace(this, EnvMissingError)
+    this.name = 'EnvMissingError'
+  }
+}
+
+export const makeValidator = <T>(parseFn: (input: string) => T, type: string = 'unknown') => {
+  return function(spec?: Spec<T>): ValidatorSpec<T> {
+    return { ...spec, type, _parse: parseFn }
+  }
+}
+
+export function bool<T extends boolean = boolean>(spec?: Spec<T>) {
+  return makeValidator((input: string | boolean) => {
+    switch (input) {
+      case true:
+      case 'true':
+      case 't':
+      case '1':
+        return true
+      case false:
+      case 'false':
+      case 'f':
+      case '0':
+        return false
+      default:
+        return null
+    }
+  }, 'bool')(spec)
+}
+
+export function num<T extends number = number>(spec?: Spec<T>) {
+  return makeValidator((input: string) => {
+    const coerced = +input
+    if (Number.isNaN(coerced)) throw new EnvError(`Invalid number input: "${input}"`)
+    return coerced
+  }, 'num')(spec)
+}
```

Comment:
Seems unnecessary that all of these returns a callback rather than `makeValidator` handling it for you, [reference](https://github.com/KATT/envsafe/blob/48fcce2beb3ade5927c073973eec70c6569ee71c/src/validators.ts#L47-L53)

Reply:
The reason for the function wrapper is to enable the `<T extends number = number>` type parameter. Check out [this PR](https://github.com/af/envalid/pull/118) for context